shell/platform/darwin/ios/framework/Source/FlutterViewController.mm

```diff
@@ -1272,65 +1284,185 @@ - (void)updateViewportPadding {
-  CGRect screenRect = [[UIScreen mainScreen] bounds];
+  self.targetViewInsetBottom = calculatedInset;
+  NSTimeInterval duration = [info[UIKeyboardAnimationDurationUserInfoKey] doubleValue];
+  [self startKeyBoardAnimation:duration];
+}
 
-  // Get the animation duration
-  NSTimeInterval duration =
-      [[info objectForKey:UIKeyboardAnimationDurationUserInfoKey] doubleValue];
+- (BOOL)shouldIgnoreKeyboardNotification:(NSNotification*)notification {
+  // Don't ignore UIKeyboardWillHideNotification notifications.
+  if (notification.name == UIKeyboardWillHideNotification) {
+    return NO;
+  }
 
-  // Considering the iPad's split keyboard, Flutter needs to check if the keyboard frame is present
-  // in the screen to see if the keyboard is visible.
-  if (CGRectIntersectsRect(keyboardFrame, screenRect)) {
-    CGFloat bottom = CGRectGetHeight(keyboardFrame);
-    CGFloat scale = [UIScreen mainScreen].scale;
-    // The keyboard is treated as an inset since we want to effectively reduce the window size by
-    // the keyboard height. The Dart side will compute a value accounting for the keyboard-consuming
-    // bottom padding.
-    self.targetViewInsetBottom = bottom * scale;
-  } else {
-    self.targetViewInsetBottom = 0;
+  // Ignore notification when keyboard's dimensions and position are all zeroes,
+  // for UIKeyboardWillChangeFrameNotification. This happens when keyboard is dragged.
+  NSDictionary* info = notification.userInfo;
+  CGRect keyboardFrame = [info[UIKeyboardFrameEndUserInfoKey] CGRectValue];
+  if (notification.name == UIKeyboardWillChangeFrameNotification &&
+      CGRectEqualToRect(keyboardFrame, CGRectZero)) {
+    return YES;
   }
-  [self startKeyBoardAnimation:duration];
-}
 
-- (void)keyboardWillBeHidden:(NSNotification*)notification {
-  NSDictionary* info = [notification userInfo];
+  // When keyboard's height/width is set to 0 by other app,
+  // do not ignore so that the inset will be set to 0.
+  if (CGRectIsEmpty(keyboardFrame)) {
+    return NO;
+  }
 
   // Ignore keyboard notifications related to other apps.
+  if ([self isKeyboardNotificationForDifferentView:notification]) {
+    return YES;
+  }
+
+  // Ignore notification if the app is not active (meaning it's running in the
+  // background, interrupted, or the app is transitioning to or from the background).
+  if (UIApplication.sharedApplication.applicationState != UIApplicationStateActive) {
+    return YES;
+  }
+
+  return NO;
+}
+
+- (BOOL)isKeyboardNotificationForDifferentView:(NSNotification*)notification {
+  NSDictionary* info = notification.userInfo;
+  // Keyboard notifications related to other apps.
   id isLocal = info[UIKeyboardIsLocalUserInfoKey];
   if (isLocal && ![isLocal boolValue]) {
-    return;
+    return YES;
   }
-
-  // Ignore keyboard notifications if engine’s viewController is not current viewController.
+  // Engine’s viewController is not current viewController.
   if ([_engine.get() viewController] != self) {
```

Comment:
does this happen when an app shows multiple flutter view controller on the same screen?

Reply:
Yeah, from my understanding this can happen when you're using add-to-app stuff like this https://github.com/flutter/flutter/issues/39036#issuecomment-528492153. This is a good point, though. I didn't test use case like that.